### PR TITLE
[datadog_cloud_configuration_rule] Implemented resource to manage cloud_configuration rules

### PR DESCRIPTION
**Pull request by symphony-elias**
Implemented new resource `datadog_cloud_configuration_rule` to manage security monitoring rules of type `cloud_configuration`.
Jira card: https://datadoghq.atlassian.net/browse/SEC-6093

### CODE REVIEW

**Comment by symphony-elias** [comment]
Tests to this specific PR pass: https://github.com/DataDog/terraform-provider-datadog/actions/runs/3741709849/jobs/6351682712#step:5:1255
Screenshot 2022-12-20 at 16 35 29

File: docs/resources/cloud_configuration_rule.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+page_title: "datadog_cloud_configuration_rule Resource - terraform-provider-datadog"
+subcategory: ""
+description: |-
+  Provides a Datadog Cloud Configuration Rule resource.
+---
+
+# datadog_cloud_configuration_rule (Resource)
+
+Provides a Datadog Cloud Configuration Rule resource.
+
```
Comment (estherk15) [inline docs/resources/cloud_configuration_rule.md:12]: Remove whitespace if it's not needed

File: docs/resources/cloud_configuration_rule.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `enabled` (Boolean) Whether the cloud configuration rule is enabled.
+- `message` (String) The message associated to the rule that will be shown in findings and signals.
+- `name` (String) The name of the cloud configuration rule.
+- `policy` (String) Policy written in rego format.
```
Comment (estherk15) [inline docs/resources/cloud_configuration_rule.md:23]:
```suggestion
- `policy` (String) Policy written in Rego format.
```

**Review by estherk15 — DISMISSED** [review]

File: go.mod [hunk 3]
```diff
@@ -1,7 +1,7 @@
 module github.com/terraform-providers/terraform-provider-datadog
 
 require (
-	github.com/DataDog/datadog-api-client-go/v2 v2.5.1-0.20221214110216-1f5a9877e971
+	github.com/DataDog/datadog-api-client-go/v2 v2.6.1
```
Comment (danieldebeer-dd) [inline go.mod:4]: Why are we updating the api client? Is this perhaps why the tests are failing?

File: docs/resources/cloud_configuration_rule.md (hunk 1 above)
Comment (symphony-elias) [reply docs/resources/cloud_configuration_rule.md:12]: Whitespaces are generated automatically from `Description` fields in file resource_datadog_cloud_configuration_rule.go methods `resourceDatadogCloudConfigurationRule` and `cloudConfigurationRuleSchema`

File: go.mod (hunk 3 above)
Comment (symphony-elias) [reply go.mod:4]: Yes and I needed the latest release with the datadog-api-spec changes which expose the rego policy and other useful fields.

**Review by estherk15 — DISMISSED** [review]

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+		return diag.FromErr(err)
+	}
+
+	if response.SecurityMonitoringStandardRuleResponse != nil {
+		d.SetId(response.SecurityMonitoringStandardRuleResponse.GetId())
+	} else {
+		return diag.FromErr(fmt.Errorf("SecurityMonitoringStandardRuleResponse is empty"))
+	}
+
+	return nil
```
Comment (skarimo) [inline datadog/resource_datadog_cloud_configuration_rule.go:123]: Need to update state here (the `updateResourceDataFromResponse` method)

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	groupByFields := getStringSlice(d, groupByField)
+
+	signalOptions := datadogV2.NewCloudConfigurationRuleComplianceSignalOptions()
+	signalOptions.SetUserActivationStatus(len(groupByFields) > 1)
+	signalOptions.SetUserGroupByFields(groupByFields)
+
+	return *signalOptions
+}
+
+func getStringSlice(d *schema.ResourceData, key string) []string {
```
Comment (skarimo) [inline datadog/resource_datadog_cloud_configuration_rule.go:164]: Lets move this function to `internal/utils/utils.go`

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+	resourceType := ruleResponse.GetOptions().ComplianceRuleOptions.ResourceType
+	regoRule := ruleResponse.GetOptions().ComplianceRuleOptions.GetRegoRule()
+	d.Set(resourceTypeField, resourceType)
+	d.Set(relatedResourceTypesField, getRelatedResourceTypes(*resourceType, regoRule.ResourceTypes))
+	d.Set(policyField, regoRule.Policy)
+
+	ruleCase := ruleResponse.GetCases()[0]
+	d.Set(severityField, ruleCase.Status)
+	d.Set(notificationsField, ruleCase.Notifications)
```
Comment (skarimo) [inline datadog/resource_datadog_cloud_configuration_rule.go:294]: For optional fields, need to check wether the field is set first. Something like:
```
	if v, ok := ruleCase.GetNotificationsOk(); ok && len(v) > 0 {
                d.Set(notificationsField, *v)
	}
```

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: cloudConfigurationRuleSchema(),
+	}
+}
+
+func cloudConfigurationRuleSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
```
Comment (skarimo) [inline datadog/resource_datadog_cloud_configuration_rule.go:41]: Looks like missing a schema for `options`. The schema should resemble the spec. Also seem to be missing `cases`?

Comment (symphony-elias) [reply datadog/resource_datadog_cloud_configuration_rule.go:41]: No, this resource only focuses on `cloud_configuration` rules and we want to expose the fields only relevant to this type of rules. `cases` and `options` are handled internally when building or parsing the json payload.

File: datadog/resource_datadog_cloud_configuration_rule.go (hunk 4 above)
Comment (symphony-elias) [reply datadog/resource_datadog_cloud_configuration_rule.go:124]: Done in line 119

File: datadog/resource_datadog_cloud_configuration_rule.go (hunk 6 above)
Comment (symphony-elias) [reply datadog/resource_datadog_cloud_configuration_rule.go]: For `cloud_configuration` rules, cases must have exactly one item with status set, see: https://github.com/DataDog/logs-backend/blob/prod/domains/cloud-security-platform/apps/security-monitoring-api/src/main/java/com/dd/logs/api/rule_engine/validators/compliance/ComplianceRuleValidator.java#L123

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Main resource type to be checked by the rule.",
+		},
+		relatedResourceTypesField: {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "Related resource types to be checked by the rule. Defaults to empty list.",
+			MinItems:    0,
+			MaxItems:    9,
```
Comment (Bit-Doctor) [inline datadog/resource_datadog_cloud_configuration_rule.go:72]: Did we decided to go with max 9 in the end? I thought it was 5 last time I followed the conversation.

File: datadog/resource_datadog_cloud_configuration_rule.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	if response.SecurityMonitoringStandardRuleResponse != nil {
+		d.SetId(response.SecurityMonitoringStandardRuleResponse.GetId())
+		updateResourceDataFromResponse(d, response.SecurityMonitoringStandardRuleResponse)
+	} else {
+		return diag.FromErr(fmt.Errorf("SecurityMonitoringStandardRuleResponse is empty"))
+	}
+
+	return nil
```
Comment (Bit-Doctor) [inline datadog/resource_datadog_cloud_configuration_rule.go:124]: I think making the error check is clearer wdyt?
```suggestion
	if response.SecurityMonitoringStandardRuleResponse == nil {
		return diag.FromErr(fmt.Errorf("SecurityMonitoringStandardRuleResponse is empty"))
	}

	d.SetId(response.SecurityMonitoringStandardRuleResponse.GetId())
	updateResourceDataFromResponse(d, response.SecurityMonitoringStandardRuleResponse)
	return nil
```

**Review by Bit-Doctor — DISMISSED** [review]

File: datadog/resource_datadog_cloud_configuration_rule.go (hunk 6 above)
Comment (Bit-Doctor) [reply datadog/resource_datadog_cloud_configuration_rule.go]: @symphony-elias `status` yes but `Notifications` is `Optional: true` as far as I can tell?

File: datadog/resource_datadog_cloud_configuration_rule.go (hunk 8 above)
Comment (symphony-elias) [reply datadog/resource_datadog_cloud_configuration_rule.go:72]: The code in the backend limits to 10 resource types (main resource type + related resource types): https://github.com/DataDog/logs-backend/blob/prod/domains/cloud-security-platform/apps/security-monitoring-api/src/main/java/com/dd/logs/api/rule_engine/validators/compliance/CloudConfigurationRuleValidator.java#L209

File: datadog/resource_datadog_cloud_configuration_rule.go (hunk 6 above)
Comment (symphony-elias) [reply datadog/resource_datadog_cloud_configuration_rule.go]: Yes but the comment was more to handle the case when `notifications` was not in the json payload of the response. I replaced `ruleCase.Notifications` with `ruleCase.GetNotifications()` which returns an empty array if the field is not present. Empty array is the default value when not specified in the TF resource.

**Review by Bit-Doctor — APPROVED** [review]

**Review by skarimo — APPROVED** [review]

**Pull request merged**